modules/container-definition/main.tf

```diff
@@ -0,0 +1,72 @@
+    dnsSearchDomains       = local.is_not_windows && length(var.dns_search_domains) > 0 ? var.dns_search_domains : null
+    dnsServers             = local.is_not_windows && length(var.dns_servers) > 0 ? var.dns_servers : null
+    dockerLabels           = length(var.docker_labels) > 0 ? var.docker_labels : null
+    dockerSecurityOptions  = length(var.docker_security_options) > 0 ? var.docker_security_options : null
+    entrypoint             = length(var.entrypoint) > 0 ? var.entrypoint : null
+    environment            = length(var.environment) > 0 ? var.environment : null
+    environmentFiles       = length(var.environment_files) > 0 ? var.environment_files : null
+    essential              = var.essential
+    extraHosts             = local.is_not_windows && length(var.extra_hosts) > 0 ? var.extra_hosts : null
+    firelensConfiguration  = length(var.firelens_configuration) > 0 ? var.firelens_configuration : null
+    healthCheck            = length(var.health_check) > 0 ? var.health_check : null
+    hostname               = var.hostname
+    image                  = var.image
+    interactive            = var.interactive
+    links                  = local.is_not_windows && length(var.links) > 0 ? var.links : null
+    linuxParameters        = local.is_not_windows && length(var.linux_parameters) > 0 ? var.linux_parameters : null
+    logConfiguration       = local.log_configuration
+    memory                 = var.memory
+    memoryReservation      = var.memory_reservation
+    mountPoints            = length(var.mount_points) > 0 ? var.mount_points : null
+    name                   = var.name
+    portMappings           = length(var.port_mappings) > 0 ? var.port_mappings : null
+    privileged             = local.is_not_windows ? var.privileged : null
+    pseudoTerminal         = var.pseudo_terminal
+    readonlyRootFilesystem = local.is_not_windows ? var.readonly_root_filesystem : null
+    repositoryCredentials  = length(var.repository_credentials) > 0 ? var.repository_credentials : null
+    resourceRequirements   = length(var.resource_requirements) > 0 ? var.resource_requirements : null
+    secrets                = length(var.secrets) > 0 ? var.secrets : null
+    startTimeout           = var.start_timeout
+    stopTimeout            = var.stop_timeout
+    systemControls         = length(var.system_controls) > 0 ? var.system_controls : null
+    ulimits                = local.is_not_windows && length(var.ulimits) > 0 ? var.ulimits : null
+    user                   = local.is_not_windows ? var.user : null
+    volumesFrom            = length(var.volumes_from) > 0 ? var.volumes_from : null
+    workingDirectory       = var.working_directory
+  }
+
+  # Strip out all null values, ECS API will provide defaults in place of null/empty values
+  container_definition = { for k, v in local.definition : k => v if v != null }
+}
+
+resource "aws_cloudwatch_log_group" "this" {
```

Comment:
There are primarily 2 ways in which logs are created/managed per container definition:
1. Via Cloudwatch logs: When you provide a name to `awslogs-group`, if the CloudWatch log group does not exists, ECS will create it. This means it is created outside of IaC and therefore will not be deleted when the task/container definition is deleted, nor tagged, etc.
2. Via Firelens: Firelens allows users to provide a configuration that will forward logs to various locations, including 3rd party locations. When using Firelens though, users will need to add a FluentBit sidecar to forward logs to the Firelens service which will send logs to the final destiation based on the `firelensConfiguration`

In this module we aim to provide support for both, but with the addition of affording users the ability to manage the Cloudwatch log group through Terraform. This is the similar scenario we face with other modules as well - RDS, Lambda, EKS - that will automatically create log groups when logging is enabled. In our modules we provide support for creating the Cloudwatch log groups so that users can tag the log groups, set the retention period, ensure logs are destroyed when the resources are, etc. This is the reason why the Cloudwatch log resource is provided in this sub-module; to allow users to control the log group through Terraform/IaC

The reason for the the added logic [here](https://github.com/clowdhaus/terraform-aws-ecs/blob/73acc1d808763ccac504eb7ddda7ceef2fc6a1ac/modules/container-definition/main.tf#L6-L16) is to support the following assumptions/defaults:
1. By default, this module enables Cloudwatch logging for the container definition because this is most common for customers. This will create the log group through Terraform to allow the log group to be tagged, a retention period set (save $), and ensure the resource are cleaned up
2. If users wish to use Firelens, the deafult Cloudwatch logging will need to be disabled (this is the reason for `enable_cloudwatch_logging`). An demonstration of this is provided in the examples:
	- [Fluentbit sidecare required for Firelens](https://github.com/clowdhaus/terraform-aws-ecs/blob/73acc1d808763ccac504eb7ddda7ceef2fc6a1ac/examples/fargate/main.tf#L78-L80)
	- [Firelens logging config that depends on fluentbit sidecar](https://github.com/clowdhaus/terraform-aws-ecs/blob/73acc1d808763ccac504eb7ddda7ceef2fc6a1ac/examples/fargate/main.tf#L100-L116)
3. Should, for some reason, users want to enable CloudWatch logging but not control it through Terraform, they can do so by simply disabling the creation of the log group through `create_cloudwatch_log_group = false` 
4. Lastly, if users do not want to enable logging at all, they can disable by setting both `enable_cloudwatch_logging` and `create_cloudwatch_log_group` to `false`